librz/core/cbin.c

```diff
@@ -4771,3 +5354,85 @@ RZ_API RzCmdStatus rz_core_bin_plugins_print(RzBin *bin, RzCmdStateOutput *state
 	rz_cmd_state_output_array_end(state);
 	return RZ_CMD_STATUS_OK;
 }
+
+static void print_arch(RzBin *bin, RzCmdStateOutput *state, int num, ut64 offset, ut64 size,
+	const char *arch, int bits, const char *machine, const char *flag, RzBinInfo *info) {
+
+	char str_fmt[30];
+	const char *fmt = "dXnss";
+
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_QUIET:
+		rz_cons_printf("%s\n", arch);
+		break;
+	case RZ_OUTPUT_MODE_JSON:
+		pj_o(state->d.pj);
+		pj_ks(state->d.pj, "arch", arch);
+		pj_ki(state->d.pj, "bits", bits);
+		pj_kn(state->d.pj, "offset", offset);
+		pj_kn(state->d.pj, "size", size);
+		if (info && !strcmp(arch, "mips")) {
+			pj_ks(state->d.pj, "isa", info->cpu);
+			pj_ks(state->d.pj, "features", info->features);
+		}
+		if (machine) {
+			pj_ks(state->d.pj, "machine", machine);
+		}
+		pj_end(state->d.pj);
+		break;
+	case RZ_OUTPUT_MODE_TABLE:
+	case RZ_OUTPUT_MODE_STANDARD:
+		if (flag && strcmp(flag, "unknown_flag")) {
+			rz_strf(str_fmt, "%s_%i %s", arch, bits, flag);
+		} else {
+			rz_strf(str_fmt, "%s_%i", arch, bits);
+		}
+		rz_table_add_rowf(state->d.t, fmt, 0, offset, size, str_fmt, machine);
+		break;
+	default:
+		rz_warn_if_reached();
+		break;
+	}
+}
+
+RZ_API void rz_core_bin_archs_print(RzBin *bin, RzCmdStateOutput *state) {
```

Comment:
Why some of the printing functions are `RZ_IPI` and this one `RZ_API`? Ideally, I think, we should not have printing functions in the API at all - it should be either passed as `char *` or more complex `RzBuffer`, etc to the API consumer, be it Rizin's own internals or some external library/bindings.

Reply:
That's because this is used also by rz-bin when doing rz-bin -A